zkir/src/v3/ir.rs

```diff
@@ -0,0 +1,643 @@
+#[tag = "ir-instruction[v3]"]
+pub enum Instruction {
+    /// Assert that `cond` has value `1`. UB if `cond` is not `0` or `1`.
+    ///
+    /// No outputs
+    Assert {
+        /// The boolean condition being asserted
+        cond: Operand,
+    },
+    /// Conditionally select a value. UB if `bit` is not `0` or `1`.
+    ///
+    /// Outputs one element, identical to `a` or `b`
+    CondSelect {
+        /// A boolean selector, if `1`, select `a`, else `b`
+        bit: Operand,
+        /// The value to select for `1`
+        a: Operand,
+        /// The value to select for `0`
+        b: Operand,
+        /// The output variable name
+        output: Identifier,
+    },
+    /// Constrains `val` to a set number of bits.
+    ///
+    /// No outputs
+    ConstrainBits {
+        /// The value to constrain
+        val: Operand,
+        /// The number of bits to constrain it to
+        bits: u32,
+    },
+    /// Constrains two values `a` and `b` to be equal.
+    ///
+    /// No outputs
+    ConstrainEq {
+        /// The first value to constrain
+        a: Operand,
+        /// The second value to constrain
+        b: Operand,
+    },
+    /// Constrains a value `val` to be a boolean (`0` or `1`).
+    ///
+    /// No outputs
+    ConstrainToBoolean {
+        /// The value to constrain
+        val: Operand,
+    },
+    /// Creates a copy of a value `val`. Superfluous, but potentially useful
+    /// in some settings, and does not extend the actual circuit.
+    ///
+    /// Outputs one element, identical to `val`
+    Copy {
+        /// The variable or immediate to copy
+        val: Operand,
+        /// The output variable name
+        output: Identifier,
+    },
+    /// Declares a value as the next public input.
+    ///
+    /// No outputs
+    DeclarePubInput {
+        /// The variable or immediate to use for the public input
+        val: Operand,
+    },
+    /// A marker informing the proof assembler that a set of preceding public
+    /// inputs belong together (typically as an instruction), and whether they
+    /// are active or not.
+    ///
+    /// Every `DeclarePubInput` should be *followed* by a `PiSkip` covering it.
+    ///
+    /// No outputs, but adds activity information to [`IrSource::prove`] and
+    /// [`IrSource::check`].
+    PiSkip {
+        /// The boolean condition under which the public input is *not* skipped
+        ///
+        /// This is only used to inform transcript processing, serving as a marker
+        /// for which public inputs comprise an instruction.
+        guard: Operand,
+        /// The number of public inputs to skip in this group
+        count: u32,
+    },
+    /// Adds two elliptic curve points. UB if either is not a valid curve point.
+    ///
+    /// Outputs 2 elements, `c_x`, `c_y`
+    EcAdd {
+        /// The affine x coordinate of `a`
+        a_x: Operand,
+        /// The affine y coordinate of `a`
+        a_y: Operand,
```

Comment:
Wasn't typed values in ZKIR one of the main things for v3? Where this should be and `a` and `b` would be the operands here.